File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        Map<Integer,Integer> map = new HashMap<>();
        int left = 0;
        long sum = 0;
        long maxy = 0;
        for(int right = 0 ; right < nums.length ; right++){
            map.put(nums[right], map.getOrDefault(nums[right],0) + 1);
            sum+= nums[right];

            while(map.get(nums[right]) > 1){
                map.put(nums[left] , map.get(nums[left]) - 1);
                sum -= nums[left];
                left++;
            }

            while(right - left + 1 > k){
                map.put(nums[left] , map.get(nums[left]) - 1);
                sum -= nums[left];
                left++;
            }

            if(right - left + 1 == k){
                maxy = Math.max(maxy , sum);
            }
        }
        return maxy;
    }
}

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution(object):
    def maximumSubarraySum(self, nums, k):
        s = set()
        l = 0
        curr_sum = 0
        ans = 0
        for r in range(len(nums)):
            while nums[r] in s:
                s.remove(nums[l])
                curr_sum -= nums[l]
                l += 1
            
            s.add(nums[r])
            curr_sum += nums[r]

            if r - l + 1 == k:
                ans = max(ans,curr_sum)
                s.remove(nums[l])
                curr_sum -= nums[l]
                l+=1
        return ans